### PR TITLE
separate R.where into two simpler functions: R.where and R.whereEq

### LINKED ISSUE #1032: simplify R.where
One of the things we often cite as a difference between Ramda and certain other utility libraries is its [simple][1] API. [`R.where`][2] is not simple (emphasis mine):

> Takes a spec object and a test object and returns true if the test satisfies the spec. Any property on the spec *that is not a function* is interpreted as an equality relation.
>
> *If the spec has a property mapped to a function, then `where` evaluates the function, passing in the test object's value for the property in question, as well as the whole test object.*

Making the behaviour of `R.where` dependent on the types of the spec's values increases the complexity of this function. The ability to provide a function is very useful, though. I propose that each of the spec's values *must* be a function. This offers several advantages:

  - simpler implementation;
  - simpler type signature;
  - simpler documentation;
  - it exposes users to the strictly more powerful functionality; and
  - it removes an edge case.

This demonstrates the edge case:

```javascript
> Array.prototype.foo = {}
{}
> Array.prototype.bar = function() {}
[Function]
> R.where({foo: Array.prototype.foo}, [])
true
> R.where({bar: Array.prototype.bar}, [])
false
```

Existing uses of `R.where` could then be trivially transformed thanks to [`R.eq`][3]:

```diff
-var spec = {x: 2};
+var spec = {x: R.eq(2)};
 R.where(spec, {w: 10, x: 2, y: 300}); //=> true
 R.where(spec, {x: 1, y: 'moo', z: true}); //=> false

 var spec2 = {x: function(val, obj) { return  val + obj.y > 10; }};
 R.where(spec2, {x: 2, y: 7}); //=> false
 R.where(spec2, {x: 3, y: 8}); //=> true

 var xs = [{x: 2, y: 1}, {x: 10, y: 2}, {x: 8, y: 3}, {x: 10, y: 4}];
-R.filter(R.where({x: 10}), xs); // ==> [{x: 10, y: 2}, {x: 10, y: 4}]
+R.filter(R.where({x: R.eq(10)}), xs); // ==> [{x: 10, y: 2}, {x: 10, y: 4}]
```

Actually, a trivial transformation is *not* possible due to another source of complexity: the function is passed the value *and the object in question* (as can be seen in the `spec2` example above). We'd need to stop passing this second argument as part of this change to make `R.eq` viable. This would lose us the ability to make decisions based on multiple fields, though this is a different problem for which `R.where` is not a good fit. The `spec2` example is better written *without* `R.where`:

```javascript
var pred = function(obj) { return obj.x + obj.y > 10; };
pred({x: 2, y: 7});  // => false
pred({x: 3, y: 8});  // => true
```

[1]: http://www.infoq.com/presentations/Simple-Made-Easy
[2]: http://ramdajs.com/docs/#where
[3]: http://ramdajs.com/docs/#eq

### PR DESCRIPTION
**Pull request by davidchambers**
Closes #1032

This is the alternative to #1034. If we decide we want *both* functions we need to decide what to name them. :)

### CODE REVIEW

File: test/where.js [hunk 1]
```diff
@@ -50,17 +27,13 @@ describe('where', function() {
   });
 
   it('is false if the test object is null-ish', function() {
```
Comment (davidchambers) [inline test/where.js]: I'd love to remove this special case.

File: src/where.js [hunk 2]
```diff
@@ -1,44 +1,46 @@
...
  */
 module.exports = _curry2(function where(spec, testObj) {
-  var parsedSpec = groupBy(function(key) {
-    return typeof spec[key] === 'function' ? 'fn' : 'obj';
-  }, keys(spec));
-
-  return _satisfiesSpec(spec, parsedSpec, testObj);
+  testObj = Object(testObj);
+  for (var prop in spec) {
+    if (_has(prop, spec) && !spec[prop](testObj[prop])) {
```
Comment (CrossEye) [inline src/where.js]: I would still rather pass the object as well.  We clearly need to hash this out.  The triangle inequality example deleted below is a reasonable example of why.

File: test/where.js (hunk 1 above)
Comment (CrossEye) [inline test/where.js]: :+1:

File: src/where.js (hunk 2 above)
Comment (davidchambers) [inline src/where.js]: My view is that multiple-field constraints are better expressed in terms of functions which take the *whole* object. The `(value, object)` parameters strike me as awkward: if one has access to the whole object, `value` isn't necessary. The counterargument is likely that `value` is convenient in the common case. To this I argue that `R.where` should focus on the common case and let multiple-field constraints be expressed in different ways.

With `R.where`:

```javascript
// valid :: { a :: Number, b :: Number, c :: Number } -> Boolean
var valid = R.where({
  a: function(a, obj) {
    return a < obj.b + obj.c;
  },
  b: function(b, obj) {
    return b < obj.a + obj.c;
  },
  c: function(c, obj) {
    return c < obj.a + obj.b;
  }
});
```

Without `R.where`:

```javascript
// valid :: { a :: Number, b :: Number, c :: Number } -> Boolean
var valid = function(obj) {
  return (obj.a < obj.b + obj.c &&
          obj.b < obj.a + obj.c &&
          obj.c < obj.a + obj.b);
};
```

Or even:

```javascript
// valid :: { a :: Number, b :: Number, c :: Number } -> Boolean
var valid = R.pipe(R.props(['a', 'b', 'c']),
                   R.sortBy(R.identity),
                   R.converge(R.lt,
                              R.last,
                              R.pipe(R.init, R.sum)));
```

File: test/where.js (hunk 1 above)
Comment (davidchambers) [inline test/where.js]: :zap:

**Comment by buzzdecafe** [comment]
IMO `where` is about specifying an interface that the input _object_ must satisfy, not that the object's individual _properties_ must satisfy _in isolation_.

**Comment by megawac** [comment]
I'm fine with this as long as `whereEq` is included

```js
whereEq = compose(where, mapObject(eq))
```

**Comment by CrossEye** [comment]
Damn, it hadn't hit me how straightforward that conversion was:

>     whereEq = compose(where, mapObject(eq))

Nice.

**Comment by davidchambers** [comment]
Shall I add `R.whereEq`?

**Comment by CrossEye** [comment]
I could be happy with this version of `where` alongside a `whereEq`.  But I believe @buzzdecafe still might have some reservations about it.

**Comment by buzzdecafe** [comment]
not reservations. just not seeing the point particularly.

**Comment by CrossEye** [comment]
I say yes, then.  Two simple functions beat one complex function hands-down.

**Comment by davidchambers** [comment]
I've added `R.whereEq`.

File: src/whereEq.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+var where = require('./where');
+
+
+/**
+ * Takes a spec object and a test object; returns true if the test satisfies
+ * the spec, false otherwise. An object satisfies the spec if, for each of the
+ * spec's own properties, accessing that property of the object gives the same
+ * value (in `R.eq` terms) as accessing that property of the spec.
+ *
+ * `whereEq` is a specialization of [`where`](#where).
```
Comment (davidchambers) [inline src/whereEq.js]: Is this valid use of the word specialization?

Comment (scott-christopher) [inline src/whereEq.js]: `whereEq` is the generalisation of `where`, while `where` is a specialisation of `whereEq`.

Comment (CrossEye) [inline src/whereEq.js]: I would say so.

**Comment by davidchambers** [comment]
((:bell:)) @buzzdecafe, @CrossEye

**Comment by buzzdecafe** [comment]
i don't mind the complexity of the original. i don't really object to this, but it doesn't excite me either. i expect this may break a lot of stuff that uses `where` at present.

this is nice: `return where(mapObj(eq, spec), testObj);`
but i wonder how it performs? that was the reason for the nastiness of the original.

**Comment by davidchambers** [comment]
> i expect this may break a lot of stuff that uses `where` at present.

Certainly. In most cases, users will need to replace `R.where` with `R.whereEq`:

```diff
-R.where({x: 1, y: 2})
+R.whereEq({x: 1, y: 2})
```

In each of the remaining cases, each non-function value in the spec must be wrapped with `R.eq`:

```diff
-R.where({x: 1, y: 2, z: R.gte(R.__, 0)})
+R.where({x: R.eq(1), y: R.eq(2), z: R.gte(R.__, 0)})
```

**Comment by CrossEye** [comment]
Sorry, I thought I'd made it clear that I was in favor of this.

I do think that sometime soon, we'll have to look at the basic implementation.  If I'm not mistaken, in the Great Curry Debacle of 2014, we reset this one in such a way that it might have lost some of the benefits that had been gained by the initial manual currying we had done.  This one was the function that showed us the need for changes to our old currying, and it had some interesting behavior...

**Comment by megawac** [comment]
Let's just make sure this is bold, highlighted, underlined in change log as
it will break a good amount of code

**Comment by buzzdecafe** [comment]
yep, totally agree

**Comment by buzzdecafe** [comment]
here is some evidence suggesting we can get a boost simply by defining `mapObjEq` in closure and using that: http://jsperf.com/mapobj-test

**Comment by buzzdecafe** [comment]
not sure what to make of this: http://jsperf.com/mapobj-test

**Comment by davidchambers** [comment]
I could certainly get behind a more efficient implementation. It was API complexity this pull request sought to reduce.

File: src/where.js [hunk 4]
```diff
@@ -1,44 +1,45 @@
...
+ *      pred({a: 'foo', b: 'xxx', x: 11, y: 20}); //=> false
  */
 module.exports = _curry2(function where(spec, testObj) {
-  var parsedSpec = groupBy(function(key) {
-    return typeof spec[key] === 'function' ? 'fn' : 'obj';
-  }, keys(spec));
-
-  return _satisfiesSpec(spec, parsedSpec, testObj);
+  for (var prop in spec) {
+    if (_has(prop, spec) && !spec[prop](testObj[prop])) {
```
Comment (megawac) [inline src/where.js]: Any reason you're not using `keys` here (it'd be faster)?

Comment (davidchambers) [inline src/where.js]: Sounds like a good idea to me!

**Pull request merged**